### PR TITLE
fix: add generator for arbitrary sized arrays

### PR DESCRIPTION
**Pull request by camshaft**
There were a few recent changes that broke the build in s2n-quic:

* In #126, the generic bounds in derives were changed and broke implementations using arbitrary sized arrays:
  ```rust
  #[derive(TypeGenerator)]
  struct Thing<T, const LEN: usize> {
    data: [T; LEN],
  }
  ```

  This is because before we generated a where clause for the whole field type:

  ```rust
  where [T; LEN]: TypeGenerator
  ```

  and now we generate it just for the generic type:

  ```rust
  where T: TypeGenerator
  ```

  To work around this, I've implemented TypeGenerator for all array lengths, which is a good idea to do anyway.
* In #109, it was set up to prioritize importing `TypeGenerator` from the `bolero` crate, if possible. This causes issues for crates like `s2n-quic-core` where we export TypeGenerator impls with a feature flag and only use `bolero` as a dev-dependency. I've swapped the priorities and this is resolved.

To try and prevent this kind of breakage in the future, I've added a workflow to test s2n-quic against any bolero changes. Note that the versions do need to match (they currently don't since the s2n-quic update is blocked on these issues) but I was able to test locally and make sure these are fixed. Moving forward, we should still be able to catch issues before bumping bolero versions.

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -137,3 +137,48 @@ jobs:
...
+
+    - name: Add patch
+      working-directory: target/integration/${{ matrix.repo }}
+      run: |
+        cat <<EOF >> Cargo.toml
+        [patch.crates-io]
+        bolero = { path = "../../../../bolero" }
+        bolero-engine = { path = "../../../../bolero-engine" }
+        bolero-generator = { path = "../../../../bolero-generator" }
+        bolero-generator-derive = { path = "../../../../bolero-generator-derive" }
```
Comment (adpaco-aws) [inline .github/workflows/main.yml:164]: Not a priority, but we could have a script that does this patching in the future.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -137,3 +137,48 @@ jobs:
...
+        EOF
+
+    - uses: actions-rs/toolchain@v1.0.7
```
Comment (adpaco-aws) [inline .github/workflows/main.yml:167]: I'd move put this action before the patching step.

File: bolero-generator/src/array.rs [hunk 3]
```diff
@@ -1,122 +1,54 @@
 use crate::{Driver, TypeGenerator, TypeGeneratorWithParams, TypeValueGenerator, ValueGenerator};
 
-impl<T> TypeGenerator for [T; 0] {
-    fn generate<D_: Driver>(_driver: &mut D_) -> Option<Self> {
-        Some([])
-    }
-}
+impl<T: TypeGenerator, const LEN: usize> TypeGenerator for [T; LEN] {
+    fn generate<D: Driver>(driver: &mut D) -> Option<Self> {
+        // TODO use core::array::try_from_fn once stable https://github.com/rust-lang/rust/issues/89379
```
Comment (adpaco-aws) [inline bolero-generator/src/array.rs:5]: Open an issue to track this enhancement, then paste it here.

File: bolero-generator/src/array.rs [hunk 4]
```diff
@@ -1,122 +1,54 @@
...
+        let mut maybe_init: [Option<T>; LEN] = [(); LEN].map(|_| None);
```
Comment (adpaco-aws) [inline bolero-generator/src/array.rs:6]: Question: Can't we just initialize to `[None; LEN]`? Or why is this needed?

File: .github/workflows/main.yml (hunk 1 above)
Comment (camshaft) [reply .github/workflows/main.yml:164]: Yeah I can open an issue for it and link it

File: bolero-generator/src/array.rs (hunk 4 above)
Comment (camshaft) [reply bolero-generator/src/array.rs:6]: No, that array literal syntax requires that the values implement `Clone`, which I don't want to force

**Review by adpaco-aws — APPROVED** [review]
Thanks, @camshaft !
